src/graph_gui.cpp

```diff
@@ -9,156 +9,184 @@
 
 #include "stdafx.h"
 #include "graph_gui.h"
-#include "window_gui.h"
 #include "company_base.h"
 #include "company_gui.h"
 #include "economy_func.h"
-#include "cargotype.h"
 #include "strings_func.h"
+#include "widget_type.h"
 #include "window_func.h"
-#include "gfx_func.h"
-#include "core/geometry_func.hpp"
 #include "currency.h"
 #include "timer/timer.h"
 #include "timer/timer_window.h"
-#include "timer/timer_game_tick.h"
-#include "timer/timer_game_calendar.h"
-#include "timer/timer_game_economy.h"
+#include "window_type.h"
 #include "zoom_func.h"
 
 #include "widgets/graph_widget.h"
 
 #include "table/strings.h"
 #include "table/sprites.h"
 
+#include "selector_gui.h"
 #include "safeguards.h"
 
-/* Bitmasks of company and cargo indices that shouldn't be drawn. */
-static CompanyMask _legend_excluded_companies;
-static CargoTypes _legend_excluded_cargo;
-
 /* Apparently these don't play well with enums. */
 static const OverflowSafeInt64 INVALID_DATAPOINT(INT64_MAX); // Value used for a datapoint that shouldn't be drawn.
 static const uint INVALID_DATAPOINT_POS = UINT_MAX;  // Used to determine if the previous point was drawn.
 
 constexpr double INT64_MAX_IN_DOUBLE = static_cast<double>(INT64_MAX - 512); ///< The biggest double that when cast to int64_t still fits in a int64_t.
 static_assert(static_cast<int64_t>(INT64_MAX_IN_DOUBLE) < INT64_MAX);
 
+/** Contains the interval of a graph's data. */
+struct ValuesInterval {
+	OverflowSafeInt64 highest; ///< Highest value of this interval. Must be zero or greater.
+	OverflowSafeInt64 lowest;  ///< Lowest value of this interval. Must be zero or less.
+};
+
+struct CargoGraphSelectorWidget : CargoSelectorWidget {
+	void OnChanged() override
+	{
+		InvalidateWindowData(WC_PAYMENT_RATES, 0);
+	}
+};
+
+struct CompanyGraphSelectorWidget : CompanySelectorWidget {
+	void OnChanged() override
+	{
+		InvalidateWindowData(WC_INCOME_GRAPH, 0);
+		InvalidateWindowData(WC_OPERATING_PROFIT, 0);
+		InvalidateWindowData(WC_DELIVERED_CARGO, 0);
+		InvalidateWindowData(WC_PERFORMANCE_HISTORY, 0);
+		InvalidateWindowData(WC_COMPANY_VALUE, 0);
+	}
+};
+
+static CompanyGraphSelectorWidget _company_selector; ///< The company selector widget, used inside #GraphLegendWindow
+static CargoGraphSelectorWidget _cargo_selector; ///< The cargo selector widget, used inside #GraphLegendWindow
+
 /****************/
 /* GRAPH LEGEND */
 /****************/
 
 struct GraphLegendWindow : Window {
+	SelectorWidget *selector; ///< The selector for the current window
+
 	GraphLegendWindow(WindowDesc &desc, WindowNumber window_number) : Window(desc)
 	{
-		this->InitNested(window_number);
-
-		for (CompanyID c = COMPANY_FIRST; c < MAX_COMPANIES; c++) {
-			if (!HasBit(_legend_excluded_companies, c)) this->LowerWidget(WID_GL_FIRST_COMPANY + c);
-
-			this->OnInvalidateData(c);
+		this->CreateNestedTree();
+		if (window_number == 0) {
```

Comment:
I think the window_numbers need to be better defined than just `== 0`

There is `WindowNumberEnum` in `window_type.h` for this purpose.